src/js/tech/middleware.js

```diff
@@ -0,0 +1,63 @@
+import Tech from './tech.js';
+import toTitleCase from '../utils/to-title-case.js';
+
+const middlewares = {};
+
+export function use(type, middleware) {
+  (middlewares[type] = middlewares[type] || []).push(middleware);
+}
+
+export function setSource(src, next) {
+  setTimeout(()=>ssh(src, middlewares[src.type] || [], next, []), 1);
+}
+
+export function setTech(middleware, tech) {
+  middleware.forEach((mw) => mw.setTech && mw.setTech(tech));
+}
+
+export const allowedGetters = {
+  currentTime: 1,
+  duration: 1
+};
+
+export const allowedSetters = {
+  setCurrentTime: 1
+};
+
+function ssh(src, middleware, next, acc) {
+  const mw = middleware[0];
+
+  // Either we've reached a leaf or we've reached a fork
+  if (typeof mw === 'string') {
+    const split = mw.split('/');
+
+    // we've reached the end, so, we need to get a tech and return.
+    if (split[0] === 'videojs') {
+      const tech = Tech.getTech(toTitleCase(split[1]));
+
+      next(tech, src, acc);
+
+    // we've reached a fork, so, we need go deeper
+    } else {
+      ssh(src, middlewares[mw] || [], next, acc);
```

Comment:
So this makes it possible to do something like `videojs.use('video/foo', 'my-standard-middleware-stack')` and when that "middleware" is hit it would then start processing the `my-standard-middleware-stack` set of middlewares, right? Seems interesting.

Although assuming that set of middleware eventually changes the type back to something standard we need to determine if we will ever call the same middleware twice.